# Python Basics

## Objective
This notebook will be used as an introduction to the basic concepts in Python.

## Free and Open-source
Python is developed under an OSI-approved open source license. Hence, it is completely free to use, even for commercial purposes. It doesn't cost anything to download Python or to include it in your application. It can also be freely modified and re-distributed. 

Python can be downloaded from the official Python website https://www.python.org/.

## Simple and expressive
It is really easy to code in Python. On the other hand, Python code can be read almost like plain English words.

Python needs to use only a few lines of code to perform complex tasks. For example, to display Hello World, you simply need to type one line `print("Hello World")`.

## Interpreted and Portable
When a programming language is interpreted, it means that the source code is executed line by line, and not all at once.Therefore, here is no need to compile Python because it is processed at runtime by the interpreter. 

Python code can be used on different machines and environments. You may write a Python code on a Mac and it will run on Windows or Linux without changes.

## Object-Oriented and Procedure-Oriented
A programming language is object-oriented if it focuses design around data and objects, rather than functions and logic. On the contrary, a programming language is procedure-oriented if it focuses more on functions (code that can be reused). One of the critical Python features is that it supports both object-oriented and procedure-oriented programming.

## Reserved words
In Python 3.6 the are 33 reserved words:

`and as assert break class continue def del elif else except False finally for from global if import in is lambda match None nonlocal not or pass raise return True try while with yield`

Remember, **Python code is case sensitive!**

## Code indentation
**Indentation** refers to the spaces at the beginning of a code line. Where in other programming languages the indentation in code is for readability only, the indentation in Python is very important. Python uses indentation to indicate a **block** of code.

## The Zen of Python
Go and check https://peps.python.org/pep-0020/

## I/O Interaction
We use the function ``print()`` to indeed _print_ the value of anything in the screen. To obtain data from the keyboard we use the method ``input()``.



In [ ]:
str = input()
print('You said:', str)

## Comments
A **comment** is a piece of text that is not executed.
In order to add a comment you must use the character # to start an the comment will end at the end of the line.

Comments are used to add relevant information to your code, notes, TODOs...

In [ ]:
# This is a COMMENT
# A comment is a piece of text that is not executed
# You can add comments to clarify your code
# TODO: explain this comment

##Variables
A **variable** is used to store data that will be used in our code. Every variable has a name that consists of letters, numbers, and the underscore character ``_``. 

Basic **data types** in Python are numbers(integers, floats), strings(texts) and booleans(true, false).

The **equal** sign = is used to assign a value to a variable. After the initial assignment is made, the value of a variable can be updated to new values and types.

We can use the `print()` to output variables (numbers, texts...) to the console. It takes one or more arguments and will output each of the arguments to the console separated by a space. If no arguments are provided, it will just output a blank line.

In [ ]:
# assign values
a_number = 10
another_number = 11.59
number_6 = 6
a_text = "this is a message to you"
another_text = 'one step beyond'

# print values
print(a_number)
print(another_number)
print(number_6)
print(a_text)
print(another_text)
print()

As you may have already noticed, *underscores* are often used for **readability**.
```
# Readable code
first_name = "Daniel"
last_name = "di la Matta"

# Somehow readable code
firstname = "Daniel"
lastname = "di la Matta"

# Non-eadable code
a = "Daniel"
b = "di la Matta"
```
Let's continue...

In [ ]:
# print several items
print("PI is rounded to", 3.14159)

In [ ]:
# oh shit!
print(unassigned_variable)

As you can see, in Python, the data type is set when you **assign** a value to a variable. The `type()` will show the data type.

In [ ]:
# reassign value and type
a = 11
print(type(a), a)
a = 11.0
print(type(a), a)
a = "11"
print(type(a), a)

##Let's play!
Create a few variables to contain the data of your new car (real or fake): 
- brand name
- model
- year
- plate number
- number of cylinders
- oil consumption in liters every 100 km

Finally, print all of them in a paragraph like this:

*My car is a [year] [brand name] [model name]. It has [number of cylinders] cylinders and it consumes [] liters every 100km. My car plate number is "[plate number]".*



In [ ]:
# Let's play: write your code here

#Operations
Let's start dealing with data.

##Operating with Numbers
Any arithmetic operation can be easily executed with Python: adding, substracting, multiplying and dividing are implemented under its symbol. Other operations such as modulus or potentiation have special symbols.

In [ ]:
print("adding 1 and 1 is ", 1 + 1)
print(5 - 2)
print(4 * 3)
print(5 / 2)
print(5 // 2)
print(5 ** 2)
print(5 ** (1/2))
print(5 % 2)

Some operations are included in the `Math` built-in module. To use them you must import the module and the place the name of the module before the function separated with a dot `.`. 

A full list of functions and constants included in the `Math` module can be found in https://www.w3schools.com/python/module_math.asp.

In [ ]:
import math

print(math.sqrt(4.0))
print(math.log(math.e ** 6.5))

Some other cool inline operations...

In [ ]:
# increment
a = 19
a += 1
print(a)

# decrement
a -= 5
print(a)

# multiply
a *= 2
print(a)

# divide
a /= 9
print(a)

##Operating with Strings
There is a lot of things that can be done with **strings** and Python provides a set of built-in methods that you can use on them. A full list of these functions can be found in https://www.w3schools.com/python/python_ref_string.asp.

Strings are **objects**. Python considers `0` as the first position in a string. Some useful string methods are the following:
- `len()` is a general purpose function that will return the size of a string.However, `count()` will return the number of occurrences of a given value within the string and `replace()` will return a string where a specified value is replaced with another specified value.
- `upper()` will convert a string into upper case while `lower()` will convert it to lower case; `swapcase()` will change the case of each character, `capitalize()` will convert the first character to capital case and `title()` will convert the first character of each word to upper case.
- `find()` and `index()` searches the string for a specified value and returns the position of where it was found while `rfind()` and `rindex()` will do the same from right to left; `startswith()` will return a boolean specifying whether a string starts with a values and `endswith()` will do the same with an ending value.

Strings can be concatenated using `+` operator and multiplied by an integer resulting in as many repetitions of its value as specified.


In [ ]:
# general purpose functions
str = "the United KingdoM"

print(len(str))
print(str.count('n'))
print(str.replace('e', 'EEE'))

In [ ]:
# case functions
str = "follow the LEADER"

print(str.capitalize())
print(str.upper())
print(str.lower())
print(str.swapcase())
print(str.title())

In [ ]:
# position functions
str = "the Mountain 5"

print(str.find('n'))
print(str.rindex('n'))
print(str.find('456'))
print(str.startswith('THE'))
print(str.endswith('5'))
print()

# but also
print('Mount' in str)
print('The' in str)

In [ ]:
# strings and numbers
str = "London and Leeds"

print(str + " are the UK")
print(str * 3)


##Operating with Booleans
Python booleans can only have two possible values:
- `True`
- `False`

They are considered **numeric** (0, 1) and can be operated either numerically or logically.

Booleans are also the result of **equalities** and **inequalities**.

You can find more information about Python booleans in https://realpython.com/python-boolean/.

In [ ]:
# numeric
print(True + True)
print(False - True)

In [ ]:
# logical
print(not True)
print(True and False)
print(True or False)
print(True ^ False)

In [ ]:
# comparissons
print(1 == 1)
print('a' == 'A')
print(1 != 1.0)
print(1 >= 3)
print('abc' <= 'cde')

##Casting
*Casting*, also known as type conversion, is a process that converts a variable's **data type** into another data type. These conversions can be **implicit** (automatically interpreted) or **explicit** (using built-in functions).

The functions ``str()``, ``int()``, ``float()`` and ``bool()`` explictly try to convert whatever the get as an argument.

In [8]:
# implicit cast
a = 2
print(a + 1.1)
print(a + True)
print()

# explicit cast
print(str(False))
print(int(2.9))
print(float("1.4"))
print(bool('False'))
print(bool(0))

3.1
3

False
2
1.4
True
False


##Let's play!
Write some Python code to ask for the birth year of the user and print back his/her age.

In [ ]:
# Let's play: write your code here

#Complex data types

## Lists
In Python, lists are ordered collections of items.

There are several ways to create a list; the most common is using the `[]` sintax. 


Python lists can contain any type of data and can be nested to build matrixes.

In [ ]:
my_number_list = [1, 2, 3, 4]
my_string_list = ['abc', "cbs", 'nbc']
my_mixed_list = [123, "hello", None]
my_2d_matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

print(my_number_list)
print(my_string_list)
print(my_mixed_list)
print(my_2d_matrix)

Python lists are indexed from 0 and negative number can be used to access the last elements in the list.

In [ ]:
my_vowels = ['a', 'e', 'i', 'o', 'u']
print(my_vowels[0])
print(my_vowels[4])
# next line is an error; uncomment and execute
# print(my_vowels[5])
print(my_vowels[-1])
print(my_vowels[-5])
# next line is an error; uncomment and execute
# print(my_vowels[-6])


Lists can be easily modified by adding new elements or removing existing ones:
- ``append()`` will add an element at the end of the list
- ``insert()`` will add an element at the specified position
- ``pop()`` will remove an element at the specified position
- ``remove()`` will remove the first element with the given value
- ``clear()`` will empty the list

Lists can be joined together using the ``+`` operator.

In [ ]:
my_list = ['Leslie', 'Karen']
print(my_list)
my_list.append('Amy')
print(my_list)
my_list.insert(1, 'Patsy')
print(my_list)
my_list.pop(2)
print(my_list)
my_list.append('Leslie')
print(my_list)
my_list.remove('Leslie')
print(my_list)
my_list.clear()
print(my_list)
print()

join = [1, 2] + [3, 4]
print(join)


Other list commonly used methods are:
- ``count()`` returns the number of mathing elements in the list 
- ``index()`` returns the index of the first element with the specified value or ``-1`` if not found
- ``reverse()`` reverses the order of the list
- ``sort()`` sorts the elements in the list


In [ ]:
primes = [1, 2, 3, 5, 7, 11]
print(primes)
print(primes.count(3))
print(len(primes))
print(primes.index(5))
primes.reverse()
print(primes)
print()

# library for radomizing objects (we will see this later)
import random
random.shuffle(primes)
print(primes)
print()

primes.sort()
print(primes)
print()

There are two special types of lists:
- Tuples, created with ``()``. Items are ordered, unchangeable, and allow duplicate values.


> ``my_tuple = ('a', 'b', 'c')``


- Sets, created with ``{}``. Items are unordered, unchangeable, and do not allow duplicate values.

> ``my_set = {'a', 'b', 'c'}``


## Dictionaries
Python dictionaries are used to store data values in `key:value` pairs. Dictionaries items are ordered, changeable and duplicated keys are not allowed.

There are several ways to create a list; the most common is using the ``{}`` sintax.

Dictionaries can contain any type of data and can be nested, meaning that in a ``key:value`` pair, a dictionary is a valid value. However, only integers, floats, strings and booleans are valid keys.

In [ ]:
my_dictionary = {
  "key1": "value1",
  "key2": 222,
  3: True,
  False: None,
  "lists": {
    "list1": [1, 2, 3],
    "list2": []
  }
}
print(my_dictionary)

Brackets ``[]`` can be used either to access a dictionary's item or to add a new ``key:value`` pair. The method ``get()`` can be used to read the value of a key. The method ``update()`` will update the dictionary with the items from the given argument.

The method ``keys()`` will return a list of all the keys in a ditionary while the method ``values()`` will return a list af all its values.

In [ ]:
my_dictionary = {
  "key1": 111,
  "key2": 222
}
print(my_dictionary)
print(my_dictionary["key1"])
print(my_dictionary.get("key2"))
print()
my_dictionary["key3"] = 333
my_dictionary["empty"] = None
print(my_dictionary)
my_dictionary.update({"empty": 0, "key1": 111111, "key10": 10})
print(my_dictionary)
print()
print(my_dictionary.keys())
print(my_dictionary.values())


There are two methods to remove items from a ditionary:
- ``pop()`` will remove the element with the specified key
- ``lear()`` will remove all the elements from the dictionary

In [ ]:
my_dictionary = {
  1: 'one',
  2: 'two',
  3: 'three'
}
print(my_dictionary)
print()
print(my_dictionary.pop(2))
print(my_dictionary)
print()
my_dictionary.clear()
print(my_dictionary)


#Code flow

##Code Indentation
**Python uses indentation to identify blocks of code.** Code within the same block should be indented at the same level.

## Conditions
The basic statement to control the flow of a Python program is the ``if`` statement. It is used along with logical conditions and boolean values.


### If statement
The complete sintax is:
```
if condition 1:
  # code 1
elif condition 2:
  # code 2
elif condition 3:
  # code 3
...
else:
  # default code
```
where ``code 1`` will be executed when ``condition 1`` is evaluated to ``True``. Otherwise, ``code 2`` will be executed when ``condition 2`` is evaluated to ``True``. Otherwise, ``code 3`` will be executed when ``condition 3`` is evaluated to ``True``... and so on. Finally, ``default code`` will be executed if all the previous conditions are evaluated to ``False``.

``elif`` and ``else`` blocks may be missing if they are not necessary in our code.

Conditions can be as simple as a boolean value, an equality or inequality, or a logical operation as explained in the *Operating with Booleans* section.

In [ ]:
# change x value to try different flows
x = 0
if x == 0:
  print(x, 'is Zero')
elif x % 2 == 0:
  print(x, 'is Even')
else:
  print(x, 'is Odd')


0 is Zero


If statements can be nested if necessary or combined along with other control flow statements.

### Match statement
With Python 3.10, when comparing just a value we can simplifly the program flow with ``match``. The sintax is
```
match expression:
  case value1:
    # code 1
  case value2:
    # code 2
  case value3 | value4 | ... | valueN:
    # code 3
  case _:
    # default code
```
where ``code 1`` will be executed if ``expression`` mathes ``value1``. If it matches ``value 2`` then ``code 2`` will be executed. If any value in the expression ``value3 | value4 | ... | valueN`` is matched then ``code 3`` will be executed. And in case none of the values above are matched, the program will execute ``default code``.

In [ ]:
# change the value of color to try different flows (it will not work since this is Python 3.7)
color = 'green'
match color:
  case 'red':
    print("It's a primary color!")
  case 'yellow':
    print("It's a primary color!")
  case 'blue':
    print("It's a primary color!")
  case 'green' | 'orange' | 'purple':
    print("It's a secondary color!")
  case _:
    print("It's a mixed color!")

### Ternary operators
Ternary operators,commonly known as conditional expressions, evaluate something based on a condition being true or not in one line.

In [ ]:
# instead of 
value = 10
if value % 2 == 0:
  value_status = "even"
else:
  value_status = "odd"

print(value, 'is', value_status)
print()

# we can write
value = 3
value_status = "even" if value % 2 == 0 else "odd"
print(value, 'is', value_status)
print()

# or we can write
value = 53
print(value, 'is', "even" if value % 2 == 0 else "odd")

## Loops
Loops are useful when we need to execute a block of code repeatedly. In Python, there are two ways to execute code in loops:
- when our code needs to be executed a fixed number of times, a ``for statement`` is what we need.
- on the other hand, when our code needs to be executed repeatedly depending on some condition, a ``while statement`` is a better choice. 

### For loops
A ``for loop`` is used to iterate over a sequence of values. These values can be stored in a List, a Dictionary, a String or a Range.

A Range is sequence of numbers with a start, an end and, optionally, a step. The sintax to create a range is

``range(start, end, step)``

In [ ]:
print(list(range(5)))
print(list(range(1,5)))
print(list(range(0, 10, 2)))

[0, 1, 2, 3, 4]
[1, 2, 3, 4]
[0, 2, 4, 6, 8]


The sintax to create a ``for loop`` is
```
for variable_name in sequence:
  # code to be repeated
else:
  # code to execute at the end
```
The ``else`` block is optional.

In [ ]:
# loop on a range
for x in range(0, 9, 3):
  print(x)

In [ ]:
# loop on a list
for x in [0, 'a', None, 7/9]:
  print(x)

In [ ]:
# loop on a dictionary
dict = {"key 1": 111, "key 2": 222, "key 3": 333}
for x in dict:
  print(x, "->", dict[x])
else:
  print("No more Data!")

There are three reserved words that we need to know to fully control loop flow:
- ``continue`` will skip the rest of the current iteration of the loop and the control flow of the program will go to the next iteration.
- ``break`` will terminate the loop and skip the ``else`` blocked if defined.
- ``pass`` will do nothing and it is usually used as a placeholder for future code.

In [ ]:
# print even numbers in the 20's and 50's on a range between 0 and 99
# uncomment print statements to understand the difference between continue and pass
for x in range(99):
  if x >= 20 and x <= 29:
    if x % 2 == 0:
      print(x)
    else:
      continue
      # print('Continue')
  elif x in range(50, 59):
    if x % 2 == 0:
      print(x)
    else:
      pass
      # print('Pass')
  elif x >= 60:
    break
else:
  print("End!")


### While loops
A ``while loop`` is used to run a block code until a certain condition is met.

The sintax to create a ``while loop`` is
```
while condition:
  # code to be repeated
  else:
  # code to execute at the end
```
The ``else`` block is optional.

The `code` inside the loop will be executed as long as ``condition`` evaluates to ``True`` and it will stop the first time it eveluates to ``False``.

As in a ``for loop``, you can use ``continue``, ``break`` and ``pass`` to skip, terminate or do nothing inside the loop.

**Be careful with your conitions!** You could cause and *infinite loop*. 

In [ ]:
# print even numbers in the 20's and 50's
# uncomment print statements to understand the continue statement and the importance of a good control flow
x = 0
while x < 99:
  if x >= 20 and x <= 29:
    if x % 2 == 0:
      print(x)
    # else:
    #   continue
  elif x in range(50, 59):
    if x % 2 == 0:
      print(x)
  x = x + 1
else:
  print("End!")

##Errors
The Python interpreter will report the errors present in your code. In fact, he interpreter will try to display the line of code where the error was detected and place a caret character `^` under the portion of the code where the error was detected.

The two most common errors are:

- `SyntaxError` will be reported when some portion of the code is **incorrect**. This can include misspelled keywords, missing or too many brackets or parentheses, incorrect operators, missing or too many quotation marks...

- `NameError` will be reported when the interpreter detects a variable that is **unknown**. This may occur when a variable is used before it has been assigned a value or if a variable name is spelled differently than the point at which it was defined.

You can find a list of Python Built-in error types in https://www.tutorialsteacher.com/python/error-types-in-python


In [ ]:
# uncomment just one line to show errors
# shoot(1)
# print(b)
# print 10

However, errors (also called **exceptions**) can be handled in our code with a `try` and `except` block. 

The complete syntax looks as follows:
```
try:
    # our code
except:
    # code to be executed when error occurs in try block
else:
    # code to be executed if try block is error-free
finally:
    # code to be executed whether an exception occured or not
```
Different error types can be handled individually. 
```
try:
    # our code
except ErrorType1:
    # code to be executed when Error Type 1 occurs in try block
except ErrorType2:
    # code to be executed when Error Type 2 occurs in try block
except:
    # code to be executed when other error occurs in try block
```
**Important**: Syntax Errors cannot be handled!

In [ ]:
try:
  a = 1
# uncomment each line to see different behaviours 
#  prin b)
#  print(b)
#  print(a)
#  print(a/0)
except NameError:
  print('A NameError exception was found but, never mind, I got it.')
except:
  print('Some error occurred but it is out of controooooool...')
else:
  print('No errors were found. Yeeeeehaw!')
finally:
  print('This is the end of the world as we know it') 

## Let's play
Write some Python code that asks the user the following questions:
```
What is the current temperature?
(C)elsius os (F)arenheit?
```
Given that 

_celsius = (farenheit - 32) * 5/9_

and 

_farenheit = celsius * 9/5 + 32_

The program should admit a ``C`` or ``c`` for Celsius and an ``F`` or ``f`` for Farenheit.

Temperature must be calculated in Fahrenheit when given in Celsius and vice versa. It must be returned in a message similar to this:
```
That's [conversion] in [converted_to]!
```
If an error occurs, the program should not crash and end up showing the fllowing message.
```
So sorry! There was an error.
```


In [ ]:
# Write your code here

# Functions

Functions are pieces of code that are defined only once to be used as many times as necessary. They are defined using the word ``def`` and they may accept parameters, providing data input to the function. Functions may also return a value using the ``return`` keyword followed by the value to be returned.

The syntax to define a function is
```
def function_name(parameter1, parameter2, ...):
  #code to be executed
  return value_to_be_returned1, value_to_be_returned2, ... 
```
where parameters and the ``return`` statement are optional.

After defined, a function needs to be *called* to be used. In order to call a function you can just write its name, followed by parentheses and the values you provide for the parameters.

In [ ]:
# let's define a function to sum two numbers
def sum(a, b):
  return a + b

# let's call the sum function; TRY different parameters and even remove one of them to see what happens
c = sum(10, 24)
print(c)

Arguments may have a default value. We need to add a ``=`` after the parameter and the default value.
 

In [ ]:
# let's define a function to print a full name
def print_name(first_name, last_name = "No surname"):
  print(first_name, last_name)

print_name("Tony", "Perkins")
print_name("Mike")

Normally, we will pass the parameters in the order they are defined. However, we can explicitly pass the values for each parameter in any order using the ``=`` operator.

In [ ]:
# let's define a function to print a full name
def print_name(first_name, last_name = "No surname"):
  print(first_name, last_name)

print_name(last_name = "Tony", first_name = "Perkins")


Our function may also have an unknown number of parameters. We just need to define th last known parameter after the ``*`` operator.

In [ ]:
# let's define to sum many numbers
def sum(*numbers):
  s = 0
  for x in numbers:
    s = s + x

  return s

print(sum(1, 2, 3, 4, 30))
print(sum(25))
print(sum())

In Python, a variable defined inside a function is called a **local** variable. It cannot be used outside of the **scope** of the function, and attempting to do so without defining the variable outside of the function will cause an error.

In [ ]:
# variable defined outside functions
my_var = 10
def print_25():
  my_var = 25
  print('Inside', my_var)

print_25()
print('Outside', my_var)

Inside 25
Outside 10


**Docstrings** are the most commonly used way to comment and document what any Python function does. A **docstring** starts and ends with three double quotes and can be single or multi line. For further information you can visit https://peps.python.org/pep-0257/.

## Lambdas
There is a special kind of functions called **lambdas** which are one-line functions and are defined as follows:

In [ ]:
# lambda definition
add = lambda x, y: x + y

print(add(3, 5))

They are quite useful when used as an argument to another function.

In [ ]:
a = [(1, 12), (5, 1), (9, 0), (12, -3)]
a.sort(key = lambda x: x[1])

print(a)

# Classes and Objects
In Python, we can create a new data types using Classes. Each entity of a Class has the same properties (attributes) and behaviours (methods).

In [ ]:
# let's define a new Class for a Person
class Person:
  first_name = "Name"
  last_name = "Surname"
  status = "single"

  def marry(self):
    self.status = "married"

  def full_name(self):
    return self.first_name + " " + self.last_name

  def shout_out(self):
    print(self.full_name())

When defining a method, the first parameter must represent the calling object itself, commonly written as ``self``.

To create a new object, we will use the name of the Class followed by parenthesis ``()``. After created, we can call the methods defined in the class and use the properties in any way.

In [ ]:
person1 = Person()
person1.first_name = "John"
person1.last_name = "Wayne"
print(person1.full_name(), 'is', person1.status)
person2 = Person()
person2.first_name = "Molly"
person2.last_name = "Mallone"
person2.marry()
print(person2.full_name(), 'is', person2.status)
print(Person.first_name, Person.last_name, 'is', Person.status)

John Wayne is single
Molly Mallone is married
Name Surname is single


Variables defined outside any method exist in any instance of the class and in the class itself. These are called *class variables*.

If we write the method ``__init(...)__`` for a class, we are defining how a newly created object is initialized. It is called every time the class is instantiated.

If we write the method ``__repr(...)__`` for a class, we are defining how a class instance is printed. 

The first parameter defined for a class method refers to the object that receives the _call_ and it is commonly named ``self``. 

In [ ]:
class Animal:
  def __init__(self, what, name, sound):
    self.what = what
    self.name = name
    self.sound = sound
  
  def __repr__(self):
    return "My name is " + self.name + " and I am a " + self.what

  def play(self):
    print(self.sound)

bob = Animal('dog', 'Bobby', 'woooofff')
lou = Animal('cat', 'Louie', 'meeeooow')
print(bob)
bob.play()
print(lou)
lou.play()

To fully understand Python as an **Object Oriented Programming Language**, we should learn about three more features:

- Inheritance
- Encapsulation
- Polymorphism

For further information you may follow this link https://www.programiz.com/python-programming/object-oriented-programming. 

# Libraries and Modules
Most Python code is written in external libraries called modules. The ``import`` statement can be used to import Python modules from other files. 

To import a full module we will write
```
import module_name
```
To import just some functionality in a module, we will write 
```
from module_name import function_list
```
You can find a complete reference on code import at https://docs.python.org/3/reference/import.html.


In [ ]:
import datetime
feb_28_1974 = datetime.date(year = 1974, month = 2, day = 28)
print(feb_28_1974)

In [ ]:
from matplotlib import pyplot as plt
plt.plot(3, 6)